src/shared/dissect-image.c

```diff
@@ -579,6 +580,28 @@ int dissect_image(
         if (!m)
                 return -ENOMEM;
 
+        r = sd_device_get_sysname(d, &sysname);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get device sysname: %m");
+        if (startswith(sysname, "loop")) {
+                const char *full_path;
+                _cleanup_free_ char *name_stripped = NULL;
+
+                r = sd_device_get_sysattr_value(d, "loop/backing_file", &full_path);
+                if (r < 0)
+                        log_debug_errno(r, "Failed to lookup image name via loop device backing file sysattr, ignoring: %m");
+
+                r = raw_strip_suffixes(basename(full_path), &name_stripped);
+                if (r < 0)
+                        return r;
+
+                free_and_replace(m->image_name, name_stripped);
+        } else {
+                r = free_and_strdup(&m->image_name, sysname);
+                if (r < 0)
+                        return r;
+        }
+
```

Comment:
validate the image name here, as suggested earlier?

Reply:
(also see next comment about this)

Reply:
ah sorry, I thought you wanted the validation at the point of use - fixed